src/main/java/com/google/sps/servlets/UserServlet.java

```diff
@@ -21,127 +24,137 @@
-          .getRequestDispatcher("/index.html");
-        dispatcher.forward(request, response);*/
+    // get form information
+    String imgURL = Jsoup.clean(request.getParameter("imgURL"), Whitelist.none());
+    String name = Jsoup.clean(request.getParameter("name"), Whitelist.none());
+    String email = Jsoup.clean(request.getParameter("email"), Whitelist.none());
+    String[] wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none()).split(",");
+    String[] teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none()).split(",");
+    String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
+
+    // add kind users to keyfactory
+    keyFactory.setKind("users");
+
+    // take the user elements and create a new entity object for it
+    FullEntity userEntity = Entity.newBuilder(keyFactory.newKey(id)).set("imgURL", imgURL).set("name", name)
+        .set("email", email)
+        .set("wishlist", Arrays.asList(wishlist).stream().map(StringValue::of).collect(Collectors.toList()))
+        .set("teachlist", Arrays.asList(teachlist).stream().map(StringValue::of).collect(Collectors.toList())).build();
+
+    // insert that new entity
+    try {
+      datastore.add(userEntity);
+    } catch (Exception DatastoreException) {
+      this.doPut(request, response);
+      return;
     }
+    response.sendRedirect("index.html");
+  }
+
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException { // update
+
+    // get form information
+    String NOTHING = Jsoup.clean(request.getParameter("NOTHING"), Whitelist.none());
+    String imgURL = Jsoup.clean(request.getParameter("imgURL"), Whitelist.none());
+    String name = Jsoup.clean(request.getParameter("name"), Whitelist.none());
+    String email = Jsoup.clean(request.getParameter("email"), Whitelist.none());
+    String wishlist = Jsoup.clean(request.getParameter("wishlist"), Whitelist.none());
+    String teachlist = Jsoup.clean(request.getParameter("teachlist"), Whitelist.none());
+    String id = Jsoup.clean(request.getParameter("id"), Whitelist.none());
+
+    // turn wishlist into a storable product for datastore
+    String[] wishlistList = wishlist.split(",");
+    List<Value<?>> wishlistValueList = new ArrayList<Value<?>>();
+
+    for (String value : wishlistList) {
+      StringValue tempStrValue = new StringValue(value);
+      wishlistValueList.add(tempStrValue);
+    }
+
+    // turn teachlist into a storable product for datastore
+    String[] teachlistList = teachlist.split(",");
+    List<Value<?>> teachlistValueList = new ArrayList<Value<?>>();
+
+    for (String value : teachlistList) {
+      StringValue tempStrValue = new StringValue(value);
+      teachlistValueList.add(tempStrValue);
+    }
+
+    // access the original entity
+    keyFactory.setKind("users");
+    Key userKey = keyFactory.newKey(id);
+    Entity user = datastore.get(userKey);
```

Comment:
FYI, doesn't make sense to get a user when you're going to **replace** it anyways.

Reply:
I wanted to make sure that previous user information wasn't replaced with empty string when the user only updates the wishlist or only updates the teachlist.

Reply:
Makes sense, though in that case you're doing _partial modifications_, in which case [`PATCH`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH) is the client should use and `doPatch()` the method that does this. For those increment updates you have two options:

1. Full user entity replace using PUT
2. Partial modifications using PATCH

For what it's worth, `doPatch` would look somewhat similar to the others except that to [update an entity](https://cloud.google.com/datastore/docs/concepts/entities#updating_an_entity) you'd first get the entity before updating it. For partial updates, just set a property if a request property was set. E.g.
```java
Entity user = datastore.get(userKey);
Entity.Builder userEntityBuilder = Entity.newBuilder(user);
if (request.getParameter("teachlist") != null) {
  String items = Jsoup.clean(request.getParameter("addTeachlistItem"), Whitelist.none()).split(",");
  userEntityBuilder = userEntityBuilder.set("teachList", items);
}
// do something similar for other request parameters
datastore.update(user);
```

Though, to unblock the other tasks from being worked on, maybe the `doPatch` stuff should be in a separate PR.

Reply:
Added more details in #22 . Don't worry about supporting partial updates in this PR.